[Style] Hide more details about the implementation of Style::SVGPaint

weinig opened the pull request:
#### e4fae8bfe7a9e0ce3f655131120391a701775eff
<pre>
[Style] Hide more details about the implementation of Style::SVGPaint
<a href="https://bugs.webkit.org/show_bug.cgi?id=299247">https://bugs.webkit.org/show_bug.cgi?id=299247</a>

Reviewed by Darin Adler.

Stop exposing the SVGPaintType enumeration and hide the details of
Style::SVGPaint behind its strong style type.

* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/values/svg/StyleSVGPaint.cpp:
* Source/WebCore/style/values/svg/StyleSVGPaint.h:

Canonical link: <a href="https://commits.webkit.org/300300@main">https://commits.webkit.org/300300@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0455956af35add8c86945d506b5143f9906cef7e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122061 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41763 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128624 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74155 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123937 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92776 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61659 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125013 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33877 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109307 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73432 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32890 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27473 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72118 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103384 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27664 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131385 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49000 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37268 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101338 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49374 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105521 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101209 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25660 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46576 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24691 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45730 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48857 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54591 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48327 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51677 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50007 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a7d24a5c86b246757b554cf2cf3a006c5dcece2e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51074

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3bab29c614b59c5ac63e48574ec8b420804198bc. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51074

darinadler commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -289,20 +289,18 @@ static bool colorIsChallengingToHighlight(const Color& color)
-            return false;
-
-        return colorIsChallengingToHighlight(style.colorResolvingCurrentColor(style.stroke().color));
+    if (fillPaint.isNone()) {
+        if (auto strokeColor = style.stroke().tryColor())
+            return colorIsChallengingToHighlight(style.colorResolvingCurrentColor(*strokeColor));
+        return false;
     }
```
What do you think of this alternative?
```suggestion
    auto color = (style.fill().isNone() ? style.stroke() : style.fill()).tryColor();
    return color && colorIsChallengingToHighlight(style.colorResolvingCurrentColor(*color));
```

darinadler commented on Source/WebCore/style/values/svg/StyleSVGPaint.cpp:
```diff
@@ -27,152 +27,103 @@
+bool containsCurrentColor(const Style::SVGPaint& paint)
 {
-    // <paint> = none | <color> | <url> [none | <color>]? 
+    if (auto color = paint.tryAnyColor())
+        return color->containsCurrentColor();
+    return false;
```
Darn, I should have suggested this too:
```suggestion
    auto color = paint.tryAnyColor();
    return color && color->containsCurrentColor();
```

weinig committed:
[Style] Hide more details about the implementation of Style::SVGPaint
https://bugs.webkit.org/show_bug.cgi?id=299247

Reviewed by Darin Adler.

Stop exposing the SVGPaintType enumeration and hide the details of
Style::SVGPaint behind its strong style type.

* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/values/svg/StyleSVGPaint.cpp:
* Source/WebCore/style/values/svg/StyleSVGPaint.h:

Canonical link: https://commits.webkit.org/300300@main
